drivers/counter/counter_ll_stm32_timer.c

```diff
@@ -0,0 +1,544 @@
+	uint32_t prescaler;
+	struct stm32_pclken pclken;
+	uint32_t irqn;
+	LOG_INSTANCE_PTR_DECLARE(log);
+};
+
+struct counter_timer_config {
+	uint32_t prescaler;
+};
+
+static inline struct counter_stm32_data *get_dev_data(
+	const struct device *dev)
+{
+	return dev->data;
+}
+
+static inline const struct counter_stm32_config *get_stm32_config(
+	const struct device *dev)
+{
+	return CONTAINER_OF(dev->config,
+			    struct counter_stm32_config, info);
+}
+
+static int start(const struct device *dev)
+{
+	TIM_TypeDef *timer = get_stm32_config(dev)->timer;
+
+	/* enable counter */
+	LL_TIM_EnableCounter(timer);
+
+	return 0;
+}
+
+static int stop(const struct device *dev)
+{
+	TIM_TypeDef *timer = get_stm32_config(dev)->timer;
+
+	/* disable counter */
+	LL_TIM_DisableCounter(timer);
+
+	return 0;
+}
+
+static uint32_t get_top_value(const struct device *dev)
+{
+	return LL_TIM_GetAutoReload(get_stm32_config(dev)->timer);
+}
+
+static uint32_t read(const struct device *dev)
+{
+	return LL_TIM_GetCounter(get_stm32_config(dev)->timer);
+}
+
+static int get_value(const struct device *dev, uint32_t *ticks)
+{
+	*ticks = read(dev);
+	return 0;
+}
+
+/* Return true if value equals 2^n - 1 */
+static inline bool is_bit_mask(uint32_t val)
+{
+	return !(val & (val + 1));
+}
+
+static uint32_t ticks_add(uint32_t val1, uint32_t val2, uint32_t top)
+{
+	uint32_t to_top;
+
+	if (likely(is_bit_mask(top))) {
+		return (val1 + val2) & top;
+	}
+
+	to_top = top - val1;
+
+	return (val2 <= to_top) ? val1 + val2 : val2 - to_top - 1;
+}
+
+static uint32_t ticks_sub(uint32_t val, uint32_t old, uint32_t top)
+{
+	if (likely(is_bit_mask(top))) {
+		return (val - old) & top;
+	}
+
+	/* if top is not 2^n-1 */
+	return (val >= old) ? (val - old) : val + top + 1 - old;
+}
+
+static void set_cc_int_pending(const struct device *dev, uint8_t chan)
+{
+	atomic_or(&get_dev_data(dev)->cc_int_pending, BIT(chan));
+	NVIC_SetPendingIRQ(get_stm32_config(dev)->irqn);
+}
+
+static int set_cc(const struct device *dev, uint8_t id, uint32_t val,
+		  uint32_t flags)
+{
+	__ASSERT_NO_MSG(get_dev_data(dev)->guard_period < get_top_value(dev));
+	bool absolute = flags & COUNTER_ALARM_CFG_ABSOLUTE;
+	bool irq_on_late;
+	TIM_TypeDef *timer = get_stm32_config(dev)->timer;
+	uint32_t top = get_top_value(dev);
+	int err = 0;
+	uint32_t prev_val;
+	uint32_t now;
+	uint32_t diff;
+	uint32_t max_rel_val;
+
+	__ASSERT(!check_it_enabled[id](timer),
+		 "Expected that CC interrupt is disabled.");
+
+	/* First take care of a risk of an event coming from CC being set to
+	 * next tick. Reconfigure CC to future (now tick is the furtherest
+	 * future).
+	 */
+	now = read(dev);
+	prev_val = get_timer_compare[id](timer);
+	set_timer_compare[id](timer, now);
+	clear_it_flag[id](timer);
+
+	if (absolute) {
+		max_rel_val = top - get_dev_data(dev)->guard_period;
+		irq_on_late = flags & COUNTER_ALARM_CFG_EXPIRE_WHEN_LATE;
+	} else {
+		/* If relative value is smaller than half of the counter range
```

Comment:
If i recall correctly the assumption was that absolute guard may not be set and half period should be safe to use. We just need something that exceeds potential longest interruption. The only case when half period would fail if half period is less than potential longest preemption but handling counter in such scenario would not make much sense.